From: John Garry <john.garry@huawei.com>
Date: Thu, 16 Apr 2020 19:18:51 +0800
Subject: [PATCH] blk-mq: Put driver tag in blk_mq_dispatch_rq_list() when no
 budget
References: bsc#1175995,jsc#SLE-15608
Git-commit: 5fe56de799ad03e92d794c7936bf363922b571df
Patch-mainline: v5.7-rc2

If in blk_mq_dispatch_rq_list() we find no budget, then we break of the
dispatch loop, but the request may keep the driver tag, evaulated
in 'nxt' in the previous loop iteration.

Fix by putting the driver tag for that request.

Reviewed-by: Ming Lei <ming.lei@redhat.com>
Signed-off-by: John Garry <john.garry@huawei.com>
Signed-off-by: Jens Axboe <axboe@kernel.dk>
Signed-off-by: Hannes Reinecke <hare@suse.de>
---
 block/blk-mq.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/block/blk-mq.c b/block/blk-mq.c
index de8e6c2f9988..b5e23e5d441e 100644
--- a/block/blk-mq.c
+++ b/block/blk-mq.c
@@ -1214,8 +1214,10 @@ bool blk_mq_dispatch_rq_list(struct request_queue *q, struct list_head *list,
 		rq = list_first_entry(list, struct request, queuelist);
 
 		hctx = rq->mq_hctx;
-		if (!got_budget && !blk_mq_get_dispatch_budget(hctx))
+		if (!got_budget && !blk_mq_get_dispatch_budget(hctx)) {
+			blk_mq_put_driver_tag(rq);
 			break;
+		}
 
 		if (!blk_mq_get_driver_tag(rq)) {
 			/*
-- 
2.16.4

